core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java

```diff
@@ -206,91 +207,83 @@ public void run() {
 
             CompletableFutures.await(stateTransferLock.topologyFuture(minTopologyId));
 
-            final List<XSiteState> chunk = new ArrayList<>(chunkSize <= 0 ? DEFAULT_CHUNK_SIZE : chunkSize);
+            final List<XSiteState> chunk = new ArrayList<>(chunkSize);
 
             if (debug) {
                log.debugf("[X-Site State Transfer - %s] start DataContainer iteration", xSiteBackup.getSiteName());
             }
 
-            for (InternalCacheEntry ice : dataContainer) {
-               if (canceled) {
-                  log.debugf("[X-Site State Transfer - %s] State transfer canceled!", xSiteBackup.getSiteName());
-                  return;
-               }
-               if (chunkSize > 0 && chunk.size() == chunkSize) {
-                  try {
-                     sendFromSharedBuffer(xSiteBackup, chunk, this);
-                  } catch (Throwable t) {
-                     error = true;
-                     log.unableToSendXSiteState(xSiteBackup.getSiteName(), t);
-                     return;
-                  }
-                  chunk.clear();
-               }
-               if (shouldSendKey(ice.getKey())) {
-                  if (trace) {
-                     log.tracef("Added key '%s' to current chunk", ice.getKey());
-                  }
-                  chunk.add(XSiteState.fromDataContainer(ice));
-               }
+            try {
+               Flowable.fromIterable(dataContainer)
+                       .filter(ice -> shouldSendKey(ice.getKey()))
+                       .map(ice -> {
+                          if (trace) {
+                             log.tracef("Added key '%s' to current chunk", ice.getKey());
+                          }
+                          return XSiteState.fromDataContainer(ice);
+                       })
+                       .buffer(chunkSize)
+                       .takeUntil(batch -> canceled)
+                       .subscribeOn(Schedulers.from(executorService))
+                       .blockingForEach(batch -> {
+                          try {
+                             // TODO Make non-blocking and use concatMapCompletable like OutboundTransferTask
+                             sendFromSharedBuffer(xSiteBackup, batch, this);
+                          } catch (Throwable t) {
+                             // This will terminate the flowable early
+                             throw new CacheException(t);
```

Comment:
Actually this is a problem now that we are using a Scheduler. It means the exception will be propagated up to the Thread uncaughtException handler. We should just log the error directly in this method instead now.

Reply:
you could just log the error and set `cancelled` to true.

Reply:
Right... I'm struggling to remember why I added a scheduler now ;)

Reply:
;)